contracts/governance/extensions/GovernorSequentialProposalId.sol

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {Governor} from "../Governor.sol";
+
+abstract contract GovernorSequentialProposalId is Governor {
+    uint256 private _proposalCount;
+    mapping(uint256 proposalHash => uint256 proposalId) private _proposalIds;
+
+    /**
+     * @dev The proposal count may only be set if the current proposal count is 0 in {_setProposalCount}.
+     */
+    error GovernorCanNotSetProposalCount();
+
+    /**
+     * @dev See {IGovernor-getProposalId}.
+     */
+    function getProposalId(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) public view virtual override returns (uint256) {
+        uint256 proposalHash = hashProposal(targets, values, calldatas, descriptionHash);
+        uint256 storedProposalId = _proposalIds[proposalHash];
+        if (storedProposalId == 0) {
+            revert GovernorNonexistentProposal(0);
+        }
+        return storedProposalId;
+    }
+
+    function _propose(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        string memory description,
+        address proposer
+    ) internal virtual override returns (uint256) {
+        uint256 proposalHash = hashProposal(targets, values, calldatas, keccak256(bytes(description)));
+        uint256 storedProposalId = _proposalIds[proposalHash];
+        if (storedProposalId == 0) {
+            _proposalIds[proposalHash] = ++_proposalCount;
+        }
+        return super._propose(targets, values, calldatas, description, proposer);
+    }
+
+    /**
+     * @dev Internal function to set the current proposal count--the next proposal id will be `newProposalCount` + 1.
```

Comment:
I don't understand this sentence.

```suggestion
     * @dev Internal function to set the current proposal count--the next proposal id will be `newProposalCount` + 1.
```

Reply:
Could you suggest a change?

Reply:
I forgot to come back and finish the suggestion. Now that I read it, it makes more sense.

I'd only suggest pointing to the definition of `proposalCount`. My confusion came because I thought the proposal count was an internal implementation detail. This should improve with some docs IMO.

```suggestion
     * @dev Internal function to set the value of {proposalCount}--the next proposal id will be `newProposalCount` + 1.
```

Reply:
I'm not a fan of the "Internal function to set the value of {proposalCount}" variant, because it disregard the parameters of `proposalCount`. Also, `proposalCount` reverts if the details don't match an existing proposal.

Therefore I think the current formulation is more correct.

Reply:
Yes good point